client/consensus/grandpa/authorities.go

```diff
@@ -160,51 +231,414 @@ func (authSet *AuthoritySet) addForcedChange(pending PendingChange, isDescendent
+				authSetChanges = append(authSetChanges, authorityChange)
+				newSet = &appliedChanges{
+					medianLastFinalized,
+					AuthoritySet{
+						currentAuthorities:     change.nextAuthorities,
+						setId:                  authSet.setId + 1,
+						pendingStandardChanges: NewChangeTree(), // new set, new changes
+						pendingForcedChanges:   nil,
+						authoritySetChanges:    authSetChanges,
+					},
+				}
+				return newSet, nil
+			}
+		}
+	}
+
+	return newSet, nil
 }
 
-// Inner Returns access to the [`AuthoritySet`].
-// TODO update with shared logic
-func (sas *SharedAuthoritySet) Inner() AuthoritySet {
-	return sas.authoritySet
+// ApplyStandardChanges Apply or prune any pending transitions based on a finality trigger. This
+// method ensures that if there are multiple changes in the same branch,
+// finalizing this block won't finalize past multiple transitions (i.e.
+// transitions must be finalized in-order). The given function
+// `is_descendent_of` should return `true` if the second hash (target) is a
+// descendent of the first hash (base).
+//
+// When the set has changed, the return value will be a Status type where newSetBlockInfo
+// is the canonical block where the set last changed (i.e. the given
+// hash and number).
+func (authSet *AuthoritySet) ApplyStandardChanges(
+	finalizedHash common.Hash,
+	finalizedNumber uint,
+	isDescendentOf IsDescendentOf,
+	initialSync bool,
+	telemetry *telemetry.Client) (Status, error) {
+	// TODO telemetry here is just a place holder, replace with real
+
+	status := Status{}
+	finalizationResult, err := authSet.pendingStandardChanges.FinalizeWithDescendentIf(&finalizedHash, finalizedNumber, isDescendentOf, applyStandardChangesPredicate(finalizedNumber))
+	if err != nil {
+		return status, err
+	}
+
+	if finalizationResult == nil {
+		return status, nil
+	}
+
+	// Changed Case
+	status.changed = true
+
+	// Flush pending forced changes to re add
+	pendingForcedChanges := authSet.pendingForcedChanges
+	authSet.pendingForcedChanges = []PendingChange{}
+
+	// we will keep all forced changes for any later blocks and that are a
+	// descendent of the finalized block (i.e. they are part of this branch).
+	for _, forcedChange := range pendingForcedChanges {
+		isDesc, err := isDescendentOf(finalizedHash, forcedChange.canonHash)
+		if err != nil {
+			return status, err
+		}
+		if forcedChange.EffectiveNumber() > finalizedNumber && isDesc {
+			authSet.pendingForcedChanges = append(authSet.pendingForcedChanges, forcedChange)
+		}
+	}
+
+	if finalizationResult.value != nil {
+		logger.Infof("%v 👴 Applying authority set change forced at block %d", initialSync, *finalizationResult.value)
+
+		// TODO add telemetry
+
+		// Store the set_id together with the last block_number for the set
+		authoritySetChange := authorityChange{
+			setId:       authSet.setId,
+			blockNumber: finalizedNumber,
+		}
+		authSet.authoritySetChanges = append(authSet.authoritySetChanges, authoritySetChange)
```

Comment:
I notice you can use `authSet.authoritySetChanges.append(authSet.setId, finalizedNumber)`

Reply:
Good call!